### ISSUE TITLE
connection error

### ISSUE THREAD

**Issue by crisward** [crystal-lang/crystal-mysql#22]
Using crystal 0.20.3 on a mac, sierra.

Getting this error

```
Error in src/app/database/database.cr:11: instantiating 'DB:Module#open(String)'

    @@db : DB::Database = DB.open(DBURL + "?max_pool_size=10&initial_pool_size=2&max_idle_pool_size=2")
                             ^~~~

in lib/db/src/db.cr:103: instantiating 'build_database(String)'

    build_database(uri)
    ^~~~~~~~~~~~~~

in lib/db/src/db.cr:117: instantiating 'build_database(URI)'

    build_database(URI.parse(connection_string))
    ^~~~~~~~~~~~~~

in lib/db/src/db.cr:121: instantiating 'DB::Database:Class#new(DB::Driver+, URI)'

    Database.new(driver_class(uri.scheme).new, uri)
             ^~~

in lib/db/src/db/database.cr:50: instantiating 'DB::Driver+#build_connection(DB::Database)'

        conn = @driver.build_connection(self).as(Connection)
                       ^~~~~~~~~~~~~~~~

in lib/mysql/src/mysql/driver.cr:3: instantiating 'MySql::Connection:Class#new(DB::Database)'

    MySql::Connection.new(db)
                      ^~~

in lib/mysql/src/mysql/connection.cr:22: instantiating 'read_packet(MySql::Protocol::HandshakeV10:Class)'

      handshake = read_packet(Protocol::HandshakeV10)
                  ^~~~~~~~~~~

in lib/mysql/src/mysql/connection.cr:57: instantiating 'read_packet()'

    read_packet do |packet|
    ^~~~~~~~~~~

in lib/mysql/src/mysql/connection.cr:47: instantiating 'build_read_packet()'

    packet = build_read_packet
             ^~~~~~~~~~~~~~~~~

in lib/mysql/src/mysql/connection.cr:65: instantiating 'MySql::ReadPacket:Class#new(TCPSocket+, MySql::Connection)'

    ReadPacket.new(@socket, self)
               ^~~

instance variable '@remaining' of MySql::ReadPacket must be Int32, not Nil

Error: instance variable '@remaining' is initialized inside a begin-rescue, so it can potentially be left uninitialized if an exception is raised and rescued
```

I can fix it by initilizing the variables inside initialize of read_packet.cr ie.

```crystal
  def initialize(@io : IO, @connection : Connection)
    @remaining = 0
    @length = 0
    @seq = 0u8
    begin
      header = uninitialized UInt8[4]
      io.read_fully(header.to_slice)
      @length = @remaining = header[0].to_i + (header[1].to_i << 8) + (header[2].to_i << 16)
      @seq = header[3]
    rescue IO::EOFError
      raise DB::ConnectionLost.new(@connection)
    end
  end
``` 
Not done a pull request as your tests appear to be passing. Not sure why mine won't compile with the same version of crystal.

**Comment by crisward** [comment]
Sorry, hadn't run `crystal deps update` since my last `git pull`, perils of working on 2 machines. It was semi-deliberate to avoid pulling the broken kemal-session-file, but that's been fixed now.
Many thanks... and good luck on the journey to crystal 1.0 in 2017.

**Comment by crisward** [comment]
Spoke too soon. Sent to my ci and got the same error. Seems to the local fix I made wasn't overwritten by a crystal deps update. For clarification here is my ci output.
Also crystal is on `Crystal 0.20.3 (2016-12-23)`

```
gitlab-ci-multi-runner 1.2.0 (3a4fcd4)
Using Shell executor...
Running on cris2.local...
Fetching changes...
Removing .shards/
Removing lib/
HEAD is now at c8e798b updated deps
Checking out c8e798bf as master...
$ crystal deps update
Updating https://github.com/kemalcr/kemal.git
Updating https://github.com/luislavena/radix.git
Updating https://github.com/jeromegn/kilt.git
Updating https://github.com/RX14/multipart.cr.git
Updating https://github.com/jeromegn/slang.git
Updating https://github.com/crystal-lang/crystal-mysql.git
Updating https://github.com/crystal-lang/crystal-db.git
Updating https://github.com/kemalcr/kemal-session-file.git
Updating https://github.com/kemalcr/kemal-session.git
Installing kemal (0.17.4)
Installing radix (0.3.5)
Installing kilt (0.3.3)
Installing multipart (0.1.1)
Installing slang (1.4.1)
Installing mysql (0.3.1)
Installing db (0.3.3)
Installing kemal-session-file (0.3.0)
Installing kemal-session (master)
$ crystal spec -v
Error in spec/spec_helper.cr:10: instantiating 'DB:Module#open(String)'

    @@db : DB::Database = DB.open(DBURL)
                             ^~~~

in lib/db/src/db.cr:103: instantiating 'build_database(String)'

    build_database(uri)
    ^~~~~~~~~~~~~~

in lib/db/src/db.cr:117: instantiating 'build_database(URI)'

    build_database(URI.parse(connection_string))
    ^~~~~~~~~~~~~~

in lib/db/src/db.cr:121: instantiating 'DB::Database:Class#new(DB::Driver+, URI)'

    Database.new(driver_class(uri.scheme).new, uri)
             ^~~

in lib/db/src/db/database.cr:50: instantiating 'DB::Driver+#build_connection(DB::Database)'

        conn = @driver.build_connection(self).as(Connection)
                       ^~~~~~~~~~~~~~~~

in lib/mysql/src/mysql/driver.cr:3: instantiating 'MySql::Connection:Class#new(DB::Database)'

    MySql::Connection.new(db)
                      ^~~

in lib/mysql/src/mysql/connection.cr:22: instantiating 'read_packet(MySql::Protocol::HandshakeV10:Class)'

      handshake = read_packet(Protocol::HandshakeV10)
                  ^~~~~~~~~~~

in lib/mysql/src/mysql/connection.cr:57: instantiating 'read_packet()'

    read_packet do |packet|
    ^~~~~~~~~~~

in lib/mysql/src/mysql/connection.cr:47: instantiating 'build_read_packet()'

    packet = build_read_packet
             ^~~~~~~~~~~~~~~~~

in lib/mysql/src/mysql/connection.cr:65: instantiating 'MySql::ReadPacket:Class#new(TCPSocket+, MySql::Connection)'

    ReadPacket.new(@socket, self)
               ^~~

instance variable '@remaining' of MySql::ReadPacket must be Int32, not Nil

Error: instance variable '@remaining' is initialized inside a begin-rescue, so it can potentially be left uninitialized if an exception is raised and rescued

ERROR: Build failed: exit status 1
```

**Comment by tbrand** [comment]
I got this as well.

**Comment by crisward** [comment]
As a temp fix, feel free to use my branch crisward/crystal-mysql for now.

**Comment by tbrand** [comment]
Thanks @criward! Will use it :+1:

**Comment by bcardiff** [comment]
I would like to be able to reproduce/isolate the issue a little bit more.

ReadPacket declaration in 0.3.1 begins as:

```
class MySql::ReadPacket
  include IO

  @length : Int32 = 0
  @remaining : Int32 = 0
  @seq : UInt8 = 0u8

  def initialize(@io : IO, @connection : Connection)
    begin
      ...
    rescue IO::EOFError
      raise DB::ConnectionLost.new(@connection)
    end
  end
```

So the ivars should not be nilable.

Can you both @tbrand @crisward  check that there is no cache `~/.crystal` / `~/.cache` bothering or wrong `./lib/crystal-mysql/src/mysql/read_packet.cr` in your path.

@tbrand what is your environment?

**Comment by crisward** [comment]
I deleted my `project/.shards` folder and removed everything in `~/.cache/crystal/`, then did a fresh `crystal deps`.

It now seems to compile again.

Bit odd it's doing it on my ci server too. I'll have a quick look into it and see if I can find the offending cache.

**Comment by crisward** [comment]
No matter what I delete, I can't get it to compile on the ci server. I've deleted every cache I can find, and it's still through the same error above. It works fine with my fork, just not the main release.

**Comment by crisward** [comment]
Just swapped my local copy back to mysql (0.3.1) from my fork, got the same error. 
Still using Crystal 0.20.3 (2016-12-23) as one of my upstream dependencies hasn't been updated to 0.20.4.

**Comment by rdp** [comment]
Make sure your crystal version is the same everywhere FWIW...0.20.5 help?

**Comment by bcardiff** [comment]
This is great! it is a bug in the compiler :-). Seriously thanks for sharing some snippets so we can found what was different between yours experiments and mine.

**Comment by bcardiff** [comment]
Since the fix in the compiler my taken a little bit longer, I've just merged #23 to unblock the class var initializers scenario.
